fabric-rendering-v1/src/client/java/net/fabricmc/fabric/api/client/rendering/v1/LayeredDrawerWrapper.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.client.rendering.v1;
+
+import java.util.function.Function;
+
+import net.minecraft.client.gui.LayeredDrawer;
+import net.minecraft.util.Identifier;
+
+/**
+ * A layered drawer that has an identifier attached to each layer and methods to add layers in specific positions.
+ *
+ * <p>Common places to add layers (as of 1.21.4):
+ * <table>
+ *     <tr>
+ *         <th>Injection Point</th>
+ *         <th>Use Case</th>
+ *     </tr>
+ *     <tr>
+ *         <td>Before {@link IdentifiedLayer#MISC_OVERLAYS MISC_OVERLAYS}</td>
+ *         <td>Render before everything</td>
+ *     </tr>
+ *     <tr>
+ *         <td>After {@link IdentifiedLayer#MISC_OVERLAYS MISC_OVERLAYS}</td>
+ *         <td>Render after misc overlays (vignette, spyglass, and powder snow) and before the crosshair</td>
+ *     </tr>
+ *     <tr>
+ *         <td>After {@link IdentifiedLayer#EXPERIENCE_LEVEL EXPERIENCE_LEVEL}</td>
+ *         <td>Render after most main hud elements like hotbar, spectator hud, status bars, experience bar, status effects overlays, and boss bar and before the sleep overlay</td>
+ *     </tr>
+ *     <tr>
+ *         <td>Before {@link IdentifiedLayer#DEMO_TIMER DEMO_TIMER}</td>
+ *         <td>Render after sleep overlay and before the demo timer, debug HUD, scoreboard, overlay message (action bar), and title and subtitle</td>
+ *     </tr>
+ *     <tr>
+ *         <td>Before {@link IdentifiedLayer#CHAT CHAT}</td>
+ *         <td>Render after the debug HUD, scoreboard, overlay message (action bar), and title and subtitle and before {@link net.minecraft.client.gui.hud.ChatHud ChatHud}, player list, and sound subtitles</td>
+ *     </tr>
+ *     <tr>
+ *         <td>After {@link IdentifiedLayer#SUBTITLES SUBTITLES}</td>
+ *         <td>Render after everything</td>
+ *     </tr>
+ * </table>
+ */
+public interface LayeredDrawerWrapper {
+	/**
+	 * Adds a layer to the end of the layered drawer.
+	 *
+	 * @param layer the layer to add
+	 * @return this layered drawer
+	 */
+	LayeredDrawerWrapper addLayer(IdentifiedLayer layer);
+
+	/**
+	 * Attaches a layer before the layer with the specified identifier.
+	 *
+	 * <p>The render condition of the layer being attached to, if any, also applies to the new layer.
```

Comment:
This is a bit unintuitive IMO, but I think there was a technical reason for this behaviour? I'd assume that people don't typically need to attach layers at specific spots anyway.

Reply:
Read the full timeline of this pull request for an explanation